prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reporter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/googleapi"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+)
+
+const reporterName = "gcsreporter"
+
+type gcsReporter struct {
+	cfg           config.Getter
+	dryRun        bool
+	logger        *logrus.Entry
+	podClientSets map[string]corev1.CoreV1Interface
+	storage       *storage.Client
+}
+
+func (gr *gcsReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	ctx := context.Background() // TODO: get this from somewhere better?
+
+	_, _, err := gr.getJobDestination(pj)
+	if err != nil {
+		return []*prowv1.ProwJob{pj}, nil
+	}
+	// TODO: do something with these errors?
+	// TODO: parallelise?
+	_ = gr.reportPodInfo(pj, ctx)
+	_ = gr.reportFinishedJob(pj, ctx)
+	_ = gr.reportProwjob(pj, ctx)
+	return []*prowv1.ProwJob{pj}, nil
+}
+
+// reportPodInfo reports information about the pod that the job ran in. In particular,
+// it reports the pod status, as well as any events concerning the pod.
+func (gr *gcsReporter) reportPodInfo(pj *prowv1.ProwJob, ctx context.Context) error {
```

Comment:
This needs to check if the ProwJob is using the Kubernetes agent and IMHO it belongs into Plank, so we don't spread agent-specific logic

Reply:
> This needs to check if the ProwJob is using the Kubernetes agent

Good point.

> IMHO it belongs into Plank, so we don't spread agent-specific logic

Also an interesting point! I originally considered putting it into plank, but put it here instead because plank is not generally supposed to be responsible for reporting. However, in so doing, I forgot that there are multiple agents and that plank is generally (but not entirely) responsible for dealing with Kubernetes - so I could see a reasonable argument for putting it back in there.